Create exit_sendmail plugin

bkabrda opened the pull request:
This PR creates `exit_sendmail` plugin - a plugin that can be used to send notifications on various conditions. There are several todo items for this PR before I'll call it ready for merge:

- [x] Document how to call the plugin
- [x] Figure out how to securely authenticate against PDC (i.e. where to get PDC auth token from)
- [x] Add a `pdc_verify_cert` argument to specify whether or not the plugin should verify PDC SSL certificate
- [x] Make the `contact_role` to send notification to configurable (now it's hardcoded as `Build_Owner` and I'm not even sure that's the correct role)
- [x] Write tests
- [x] Figure out how to get name of user who submitted the build, assuming we're notifying after a user-submitted build (low-prio, this will most likely be done by Koji when deployed in OSBS => low prio, probably not a blocker for merging)

When this is done, we'll need to implement setting the PDC source secret in osbs-client. This can be implemented along with enabling the sendmail plugin.

bkabrda commented:
Also, this depends on #280

twaugh commented on atomic_reactor/plugins/exit_sendmail.py:
```diff
@@ -0,0 +1,207 @@
+class SendMailPlugin(ExitPlugin):
+    key = "sendmail"
+    allowed_states = set(['manual_success', 'manual_fail', 'auto_success', 'auto_fail',
+                          'auto_canceled'])
```
It would be nice to have symbolic constants for these strings, like

```
MANUAL_SUCCESS = 'manual_success'
...
allowed_states = set([self.MANUAL_SUCCESS, ...])
```

That way typos will be caught more quickly.

bkabrda replied:
Good idea, will do.

bkabrda commented:
So it seems that we could have a file with PDC token on the host machine and then use the approach described at https://github.com/projectatomic/atomic-reactor/issues/270 to mount it in, read it and use it for authentication. At least that's the way I understand it, can you please confirm, @twaugh?

twaugh commented:
Yes, that's correct.

The osbs-client specifies the name of the secret and the path it will be mounted at, so the plugin will need to accept a path in its configuration so it knows where to look.

bkabrda commented:
This PR is now ready for review. I'm working on osbs-client related changes right now.

twaugh commented on atomic_reactor/plugins/exit_sendmail.py:
```diff
@@ -0,0 +1,263 @@
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param url: URL to OSv3 instance where the build logs are stored
+        :param send_on: list of build states when a notification should be sent
```
Would be nice to list `url` and `send_on` in the order they appear in the keyword list.

bkabrda replied:
Good catch, will do :)

twaugh commented on atomic_reactor/plugins/exit_sendmail.py:
```diff
@@ -0,0 +1,263 @@
+        self.error_addresses = error_addresses
+
+    def _should_send(self, rebuild, success, canceled):
+        """Return True if mail should be sent under given conditions and `self.send_on`."""
```
Should it be "in `self.send_on`"?

bkabrda replied:
Ugh, the docstring doesn't make much sense. I'll rewrite it.

twaugh commented on atomic_reactor/plugins/exit_sendmail.py:
```diff
@@ -0,0 +1,263 @@
+        if self.pdc_secret_path is not None:
+            token_file = os.path.join(self.pdc_secret_path, self.PDC_TOKEN_FILE)
+        else:
+            token_file = os.path.join(os.environ['SOURCE_SECRET_PATH'], self.PDC_TOKEN_FILE)
```
The `sourceSecret` slot is already taken by pulp, so it's `pdc_secret_path` or nothing.

bkabrda replied:
And I always took pride in not doing C&P mistakes ;)

twaugh commented on atomic_reactor/plugins/exit_sendmail.py:
```diff
@@ -0,0 +1,263 @@
+                    body
+                ])
+                receivers = self.error_addresses
+            self.log.info('sending notification to %s ...' % receivers)
```
`receivers` should be its own argument instead of using '%' here.

bkabrda replied:
Thanks, I missed that.

twaugh commented:
Looks good aside from the comments above and the TODOs already noted.

jpopelka commented:
LGTM

bkabrda commented:
I fixed the issues discovered and also added a `submitter` (user who created the build) argument to the plugin; I think this is the best way to go - osbs-client will re-set the submitter on every submitted build and autorebuilds have special value `<autorebuild>`.

bkabrda commented:
Everything is now fixed and implemented, ready to be pulled.

twaugh commented:
LGTM
